### DISCUSSION TITLE
OpenBSD port request

### DISCUSSION THREAD

**Discussion by Scorpox59** [odin-lang/Odin#991] [Ideas/Requests]
Hi gingerBill,
I would like to try Odin on OpenBSD, but it doesn't compile.

clang src/main.cpp -Wno-switch -Wno-pointer-sign -Wno-tautological-constant-out-of-range-compare -Wno-tautological-compare -Wno-macro-redefined -std=c++14 -DGIT_SHA=\"\" -DODIN_VERSION_RAW=\"dev-2021-05\" -g -pthread -ldl -lm -lstdc++ -o odin
In file included from src/main.cpp:3:
In file included from src/common.cpp:19:
src/gb/gb.h:201:4: error: This UNIX operating system is not supported
                #error This UNIX operating system is not supported
                 ^
src/gb/gb.h:321:12: fatal error: 'sys/sendfile.h' file not found
                #include <sys/sendfile.h>
                         ^~~~~~~~~~~~~~~~

**Comment by Platin21** [comment]
There was once a pull request for this you might want to take a look at that. 
This is it #745

**Comment by ghost** [comment]
I did some of the FreeBSD port quite a while ago, and I was working on NetBSD but never finished it. I can't remember everything that I did for FreeBSD, but I think you need to add defines, you might have to change the src/gb/gb.h file. You will also need to add targets to src/build_settings.cpp, I think. And then add support to the core library (which required looking up how some things are defined in FreeBSD).

Although, much of this work for FreeBSD was done *prior* to the new backend, so I'm not sure what has changed - I'm not even sure if FreeBSD works anymore since the switch :D 
But you can see what I did for FreeBSD Here: https://github.com/odin-lang/Odin/pull/734

**Comment by flashcardsmobile-admin** [comment]
I would definitely like some more OpenBSD support. When running `make release-native`, it reads:

<pre>
./build_odin.sh release-native
+ clang++ src/main.cpp src/libtommath.cpp -Wno-switch -Wno-macro-redefined -Wno-unused-value -DODIN_VERSION_RAW="dev-2026-05" -std=c++14 -I/usr/local/include -I/usr/include -std=c++17 -fno-exceptions -fno-rtti -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -L/usr/lib -O3 -march=native -pthread -lm -lstdc++ -L/usr/local/lib -liconv -lLLVM -o odin
In file included from src/main.cpp:80:
src/llvm_backend.cpp:3042:3: error: use of undeclared identifier 'LLVMInitializeAArch64TargetInfo'
 3042 |                 LLVMInitializeAArch64TargetInfo();
      |                 ^
src/llvm_backend.cpp:3043:3: error: use of undeclared identifier 'LLVMInitializeAArch64Target'
 3043 |                 LLVMInitializeAArch64Target();
      |                 ^
src/llvm_backend.cpp:3044:3: error: use of undeclared identifier 'LLVMInitializeAArch64TargetMC'
 3044 |                 LLVMInitializeAArch64TargetMC();
      |                 ^
src/llvm_backend.cpp:3045:3: error: use of undeclared identifier 'LLVMInitializeAArch64AsmPrinter'
 3045 |                 LLVMInitializeAArch64AsmPrinter();
      |                 ^
src/llvm_backend.cpp:3046:3: error: use of undeclared identifier 'LLVMInitializeAArch64AsmParser'
 3046 |                 LLVMInitializeAArch64AsmParser();
      |                 ^
src/llvm_backend.cpp:3047:3: error: use of undeclared identifier 'LLVMInitializeAArch64Disassembler'
 3047 |                 LLVMInitializeAArch64Disassembler();
      |                 ^
src/llvm_backend.cpp:3051:3: error: use of undeclared identifier 'LLVMInitializeWebAssemblyTargetInfo'
 3051 |                 LLVMInitializeWebAssemblyTargetInfo();
      |                 ^
src/llvm_backend.cpp:3052:3: error: use of undeclared identifier 'LLVMInitializeWebAssemblyTarget'
 3052 |                 LLVMInitializeWebAssemblyTarget();
      |                 ^
src/llvm_backend.cpp:3053:3: error: use of undeclared identifier 'LLVMInitializeWebAssemblyTargetMC'
 3053 |                 LLVMInitializeWebAssemblyTargetMC();
      |                 ^
src/llvm_backend.cpp:3054:3: error: use of undeclared identifier 'LLVMInitializeWebAssemblyAsmPrinter'
 3054 |                 LLVMInitializeWebAssemblyAsmPrinter();
      |                 ^
src/llvm_backend.cpp:3055:3: error: use of undeclared identifier 'LLVMInitializeWebAssemblyAsmParser'
 3055 |                 LLVMInitializeWebAssemblyAsmParser();
      |                 ^
src/llvm_backend.cpp:3056:3: error: use of undeclared identifier 'LLVMInitializeWebAssemblyDisassembler'
 3056 |                 LLVMInitializeWebAssemblyDisassembler();
      |                 ^
src/llvm_backend.cpp:3059:3: error: use of undeclared identifier 'LLVMInitializeRISCVTargetInfo'; did you mean 'LLVMInitializeX86TargetInfo'?
 3059 |                 LLVMInitializeRISCVTargetInfo();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                 LLVMInitializeX86TargetInfo
/usr/include/llvm/Config/Targets.def:2:1: note: 'LLVMInitializeX86TargetInfo' declared here
    2 | LLVM_TARGET(X86)
      | ^
/usr/include/llvm-c/Target.h:42:8: note: expanded from macro 'LLVM_TARGET'
   42 |   void LLVMInitialize##TargetName##TargetInfo(void);
      |        ^
<scratch space>:129:1: note: expanded from here
  129 | LLVMInitializeX86TargetInfo
      | ^
In file included from src/main.cpp:80:
src/llvm_backend.cpp:3060:3: error: use of undeclared identifier 'LLVMInitializeRISCVTarget'; did you mean 'LLVMInitializeX86Target'?
 3060 |                 LLVMInitializeRISCVTarget();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                 LLVMInitializeX86Target
/usr/include/llvm/Config/Targets.def:2:1: note: 'LLVMInitializeX86Target' declared here
    2 | LLVM_TARGET(X86)
      | ^
/usr/include/llvm-c/Target.h:46:38: note: expanded from macro 'LLVM_TARGET'
   46 | #define LLVM_TARGET(TargetName) void LLVMInitialize##TargetName##Target(void);
      |                                      ^
<scratch space>:133:1: note: expanded from here
  133 | LLVMInitializeX86Target
      | ^
In file included from src/main.cpp:80:
src/llvm_backend.cpp:3061:3: error: use of undeclared identifier 'LLVMInitializeRISCVTargetMC'; did you mean 'LLVMInitializeX86TargetMC'?
 3061 |                 LLVMInitializeRISCVTargetMC();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                 LLVMInitializeX86TargetMC
/usr/include/llvm/Config/Targets.def:2:1: note: 'LLVMInitializeX86TargetMC' declared here
    2 | LLVM_TARGET(X86)
      | ^
/usr/include/llvm-c/Target.h:51:8: note: expanded from macro 'LLVM_TARGET'
   51 |   void LLVMInitialize##TargetName##TargetMC(void);
      |        ^
<scratch space>:137:1: note: expanded from here
  137 | LLVMInitializeX86TargetMC
      | ^
In file included from src/main.cpp:80:
src/llvm_backend.cpp:3062:3: error: use of undeclared identifier 'LLVMInitializeRISCVAsmPrinter'; did you mean 'LLVMInitializeX86AsmPrinter'?
 3062 |                 LLVMInitializeRISCVAsmPrinter();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                 LLVMInitializeX86AsmPrinter
/usr/include/llvm/Config/AsmPrinters.def:2:1: note: 'LLVMInitializeX86AsmPrinter' declared here
    2 | LLVM_ASM_PRINTER(X86)
      | ^
/usr/include/llvm-c/Target.h:57:8: note: expanded from macro 'LLVM_ASM_PRINTER'
   57 |   void LLVMInitialize##TargetName##AsmPrinter(void);
      |        ^
<scratch space>:141:1: note: expanded from here
  141 | LLVMInitializeX86AsmPrinter
      | ^
In file included from src/main.cpp:80:
src/llvm_backend.cpp:3063:3: error: use of undeclared identifier 'LLVMInitializeRISCVAsmParser'; did you mean 'LLVMInitializeX86AsmParser'?
 3063 |                 LLVMInitializeRISCVAsmParser();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                 LLVMInitializeX86AsmParser
/usr/include/llvm/Config/AsmParsers.def:2:1: note: 'LLVMInitializeX86AsmParser' declared here
    2 | LLVM_ASM_PARSER(X86)
      | ^
/usr/include/llvm-c/Target.h:63:8: note: expanded from macro 'LLVM_ASM_PARSER'
   63 |   void LLVMInitialize##TargetName##AsmParser(void);
      |        ^
<scratch space>:145:1: note: expanded from here
  145 | LLVMInitializeX86AsmParser
      | ^
In file included from src/main.cpp:80:
src/llvm_backend.cpp:3064:3: error: use of undeclared identifier 'LLVMInitializeRISCVDisassembler'; did you mean 'LLVMInitializeX86Disassembler'?
 3064 |                 LLVMInitializeRISCVDisassembler();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                 LLVMInitializeX86Disassembler
/usr/include/llvm/Config/Disassemblers.def:2:1: note: 'LLVMInitializeX86Disassembler' declared here
    2 | LLVM_DISASSEMBLER(X86)
      | ^
/usr/include/llvm-c/Target.h:69:8: note: expanded from macro 'LLVM_DISASSEMBLER'
   69 |   void LLVMInitialize##TargetName##Disassembler(void);
      |        ^
<scratch space>:149:1: note: expanded from here
  149 | LLVMInitializeX86Disassembler
      | ^
In file included from src/main.cpp:80:
src/llvm_backend.cpp:3067:3: error: use of undeclared identifier 'LLVMInitializeARMTargetInfo'
 3067 |                 LLVMInitializeARMTargetInfo();
      |                 ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
*** Error 1 in /home/jackson/builds/Odin-dev-2026-05 (Makefile:19 'release-native')
</pre>

**Reply by heavyrain266** [reply]
I will look into this tonight.

**Reply by heavyrain266** [reply]
Could you give me the output of `llvm-config --targets-built`? It seems that you're running heavily stripped down LLVM bundled with base system (doesn't support cross-compilation)

**Reply by flashcardsmobile-admin** [reply]
It gave me: `X86 AMDGPU`

**Reply by heavyrain266** [reply]
> It gave me: `X86 AMDGPU`

You're attempting to compile it with stripped down version that doesn't support cross-compilation required to build the compiler. Try `pkg_add llvm-21.1.8p4`, it "works" for me, except that there is a dynamic linking issue for `libLLVM-0.0.so` which I'm unable to resolve yet.
